### ISSUE TITLE
git may need to be marked unthreadable (for "git commit")

### ISSUE THREAD

**Issue by mqualizz** [xonsh/xonsh#3263]
I believe this has been an issue since 0.9.9 (I may have missed 0.9.8, so I'm not sure), but running "git commit" results in:

BlockingIOError: [Errno 35] write could not complete without blocking

## xonfig
```
+------------------+----------------------+
| xonsh            | 0.9.9                |
| Git SHA          | 5aac8711             |
| Commit Date      | Jul 19 17:43:33 2019 |
| Python           | 3.7.4                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 2.0.9                |
| shell type       | prompt_toolkit2      |
| pygments         | 2.4.2                |
| on posix         | True                 |
| on linux         | False                |
| on darwin        | True                 |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
```

## Expected Behavior
Ideally, the git comment editor should appear.

## Current Behavior
Things blow up.  My instance of iTerm becomes sullen and refuses to talk to me.  And I *think* I saw my Mac giving my dog an angry look, but that could just be my imagination.

### Traceback (if applicable)

```
Exception in thread Thread-114:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 14206, in run
    i = self._read_write(procout, stdout, sys.__stdout__)
  File "/usr/local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 14250, in _read_write
    stdbuf.flush()
```

## Steps to Reproduce
$ mkdir tmp
$ cd tmp
$ git init .
$ touch foo
$ git add foo
$ git commit

**Comment by scopatz** [comment]
Hi there @mqualizz - thanks for reporting! This works fine for me.  What is your `$EDITOR` set to?

**Comment by mqualizz** [comment]
Thanks for looking, @scopatz.  $EDITOR == "nvim" -- I verified that if I change it to "vim" then there is no issue.

**Comment by gforsyth** [comment]
That's very strange.  Also works on my end using `vim` as my `$EDITOR` -- but we've already listed `nvim` as unthreadable so I'm not sure why it's causing an issue.

**Comment by deeuu** [comment]
Can confirm on `0.9.11`: No problem with `vim`, `BlockingIOError` with `nvim`.

**Comment by gotgenes** [comment]
I encountered this yesterday, also on Mac OS (10.14.6 Mojave), also with `$EDITOR = 'nvim'`

```
$ xonfig                                                                                                                
+------------------+-----------------+
| xonsh            | 0.9.13          |
| Python           | 3.8.1           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 2.0.10          |
| shell type       | prompt_toolkit2 |
| pygments         | 2.4.2           |
| on posix         | True            |
| on linux         | False           |
| on darwin        | True            |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```

I'm wondering what's a recommended workaround if we want to invoke an editor with `git commit` and not pass the message via `-m`.

**Comment by gforsyth** [comment]
> I'm wondering what's a recommended workaround if we want to invoke an editor with `git commit` and not pass the message via `-m`.

The workaround is to use `vim` for commit messages (if you want a `vim`my experience).  I'm going to try to figure out what it is about `nvim` that's causing trouble.

**Comment by deeuu** [comment]
Yeah for the moment I have
```
[core]
    editor = vim
```
in my `~/.gitconfig`

**Comment by gotgenes** [comment]
I found running as `$[git commit]` to be a non-invasive workaround suitable for now.

**Comment by pbutler** [comment]
I see this too when sshing into linux hosts from an OSX machine @gotgenes fix works for me but it happens in both vim and neovim.

**Comment by pbutler** [comment]
Might be unrelated but running neovim through xonsh (again on a linux host) also seems to induce neovim to create a directory `./~/.cache/nvim` and sometimes a log file `.nvimlog`.   The `$[]` trick doesn't work to prevent that.  This doesn't happen on bash.  Something weird is  definitely going on...

**Comment by laloch** [comment]
> Might be unrelated but running neovim through xonsh (again on a linux host) also seems to induce neovim to create a directory ./~/.cache/nvim and sometimes a log file .nvimlog.

Hi @pbutler, what does `echo $HOME` say? The above sounds like your `$HOME` environment variable is set to literal "~".

**Comment by pbutler** [comment]
Yeah

> > Might be unrelated but running neovim through xonsh (again on a linux host) also seems to induce neovim to create a directory ./~/.cache/nvim and sometimes a log file .nvimlog.
> 
> Hi @pbutler, what does `echo $HOME` say? The above sounds like your `$HOME` environment variable is set to literal "~".

Yeah.  Sorry. I spent a bit of time tracking it down. it was a bad variable setting.  I had played with `XDG_DATA_HOME` in `.xonshrc` to diagnose something else and that was causing it.  The only problem that is xonsh related is the original issue.

**Comment by gotgenes** [comment]
I'm still experiencing this with xonsh 0.9.17. I'm now at least getting a traceback:

```
                                                                               Exception in thread Thread-60:
Traceback (most recent call last):
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/xonsh/0.9.17/libexec/lib/python3.8/site-packages/xonsh/proc.py", line 662, in run
    i = self._read_write(procout, stdout, sys.__stdout__)
  File "/usr/local/Cellar/xonsh/0.9.17/libexec/lib/python3.8/site-packages/xonsh/proc.py", line 706, in _read_write
    stdbuf.flush()
BlockingIOError: [Errno 35] write could not complete without blocking
```

It doesn't look particularly helpful, but it's something.

```
$ xonfig info                                                                                                           
+------------------+-----------------+
| xonsh            | 0.9.17          |
| Python           | 3.8.2           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 3.0.4           |
| shell type       | prompt_toolkit2 |
| pygments         | 2.6.1           |
| on posix         | True            |
| on linux         | False           |
| on darwin        | True            |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```

**Comment by gotgenes** [comment]
@gforsyth Do you have an outline of what you would try to in order to figure out why the `BlockingIOError` happens with `nvim` as the editor but not with `vim`?

**Comment by gforsyth** [comment]
hey @gotgenes -- all I have right now is to confirm that `neovim` and/or `nvim` are showing up marked as "False" or similar in `__xonsh__.command_cache.threadable_predictors`.  After that, I'm sort of out of ideas.

**Comment by anki-code** [comment]
Closed with reference to #4214
